[RFC] third-party: build all deps with debugging symbols

mhinz opened the pull request:
By default, our Makefile provides `-DCMAKE_BUILD_TYPE` to `CMakeLists.txt`, but not to `third-party/CMakeLists.txt`.

Currently this is only done for unibilium, but I guess it makes sense to make debug builds of all our dependencies if `CMAKE_BUILD_TYPE=Debug` anyway.

This makes it easy to debug the bundled deps in a debugger, since you have access to the C sources.

justinmk commented on third-party/CMakeLists.txt:
```diff
@@ -11,6 +11,12 @@ if(NOT CMAKE_BUILD_TYPE)
+if(CMAKE_BUILD_TYPE MATCHES Debug)
+  set(DEBUG 1)
+else()
+  set(DEBUG 0)
```
Isn't this already set by now?

mhinz replied:
`nvim` (`CMakeLists.txt`) and `NVIM_DEPS` (`third-party/CMakeLists.txt`) are two separate cmake projects with separate caches. Since `-DCMAKE_BUILD_TYPE` is never provided to `third-party/CMakeLists.txt`, `DEBUG` isn't set as well.

The only code that sets `CMAKE_BUILD_TYPE` in `NVIM_DEPS` so far is:

```cmake
# In Windows/MSVC CMAKE_BUILD_TYPE changes the paths/linking of the build
# recipes (libuv, msgpack), make sure it is set
if(NOT CMAKE_BUILD_TYPE)
  set(CMAKE_BUILD_TYPE Release)
endif()
```

justinmk replied:
`DEBUG` is specific to unibilium, so this check should live in `third-party/cmake/BuildUnibilium.cmake`.

justinmk commented:
Before now .deps always was in a predictable state (I assume). This will change that, .deps will be in whatever state it was first built ...

mhinz commented:
I really just want _any_ way to build a proper debug environment. Currently that's not possible for the bundled deps. I figured it would make sense to make a debug build of _all_ deps (eventually) when `make CMAKE_BUILD_TYPE=Debug` is run.

Alternatively I can change the PR to only keep the one-line change in `third-party/cmake/BuildUnibilium.cmake`, then one could at least make a debug build of unibilium by running `make DEPS_CMAKE_FLAGS="-DDEBUG=1"`.

justinmk commented:
> Currently that's not possible for the bundled deps. I figured it would make sense to make a debug build of all deps (eventually) when make (since CMAKE_BUILD_TYPE=Debug the default) is run.

Yeah, there will be a "transition time" for developers who aren't used to that, but I think you're right.

@bfredl had some similar work for libuv IIRC, let's throw that in as well.

And let's do this _only_ for `CMAKE_BUILD_TYPE=Debug` no other build types. Also need:

- mention it somewhere (briefly), e.g. `src/nvim/README.md`
- maybe some way to warn when switching to non-debug build type, that  `.deps` needs to be removed. I would guess when someone does `CMAKE_BUILD_TYPE=Release` they would be unhappy if `.deps` were still non-optimized.

bfredl commented:
> @bfredl had some similar work for libuv

It was for building libuv with ASAN. It built locally but I couldn't get it to work on travis.

mhinz commented:
> And let's do this only for CMAKE_BUILD_TYPE=Debug no other build types.

Yup, that's how it works right now.

> mention it somewhere (briefly), e.g. src/nvim/README.md

You know proper wording is not my strength! :> How about adding a small section `Build notes`:

_If you run a Debug build with the bundled third-party libraries (`USE_BUNDLED=1` and `CMAKE_BUILD_TYPE=Debug`, both defaults), then these third-party libraries will build with debug information as well. Very useful when using gdb/lldb to step into library functions._

> maybe some way to warn when switching to non-debug build type

That's a bit tricky. I did this..

- `make distclean`, `make`: Both caches have `Debug` type now.
- `rm -r build`, `make CMAKE_BUILD_TYPE=Release`: Unibilium does _not_ get rebuilt, only reinstalled, but both caches say `Release` now.

So.. I can't think of any other way to detect a difference of type between core and third-party deps.

justinmk commented:
The libuv readme implies that "debug" is the default build-type:

https://github.com/libuv/libuv/blob/cf6d047e8455e354689a5ebad7d677168eee02fb/README.md#unix

AFAICT we don't specify that in `third-party/cmake/BuildLibuv.cmake`, I wonder if our bundled libuv is built in debug-mode.

More generally this raises the point that our bundled deps are not built in a consistent way. So it would be good to be more explicit about it.

mhinz commented:
Okay, I'll have a look at all the compile options (and their defaults) for our dependencies and post it here soon.

jamessan commented:
> The libuv readme implies that "debug" is the default build-type:

I thought that was obvious from the fact that we've had users hit libuv asserts...

justinmk commented:
@jamessan True, but I'm wondering if the libuv default-debug build enables any optimizations.  IOW is it closer to the Nvim "Debug" build or the Nvim "RelWithDebInfo" build?  (I haven't checked)

mhinz commented:
The following table contains the `-g` and `-O` options. `Defaults` are the, well, defaults and what Neovim uses. The `Debug` section says what we have to change to get a proper debug build.

This following table is true for `CMAKE_BUILD_TYPE=Debug` (since some of the deps use cmake as well).

| Dep | Defaults | Debug |
|-----|----------|--------|
| unibilium | `-O2` | `make CFLAGS=-O0 DEBUG=1` |
| msgpack | `-g -O3` | `cmake . -DCMAKE_C_FLAGS_DEBUG="-O0 -ggdb"` |
| libuv | `-g -O2` | `./configure CFLAGS="-O0 -ggdb"` |
| luv | `-g -O2` | `cmake . -DCMAKE_C_FLAGS_DEBUG="-O0 -ggdb"` |
| libvterm | | `make CFLAGS=-O0 DEBUG=1` |
| libtermkey |  | `make CFLAGS=-O0 DEBUG=1` |
| jemalloc | `-g3 -O3` | `./configure CFLAGS="-O0 -ggdb"` |
| gperf | `-g -O2` | `./configure CXXFLAGS="-O0 -ggdb"` |
| luajit | `-g -O2` | haven't checked yet |

Should I change the `third-party/cmake/Build*.cmake` for all of them?

justinmk commented:
@mhinz  Nice, can you include that table in the commit message?

> Should I change the third-party/cmake/Build*.cmake for all of them? Most sport -g already, maybe that's good enough. But most have optimisations enabled nevertheless.

Yes, that's fine.

> The only deps without -g flag so far are: unibilium, libvterm, libtermkey. Let's enable -g3 for these only and get this PR merged

:+1:

justinmk commented:
@mhinz  Nice, can you include that table in the commit message?

> Should I change the third-party/cmake/Build*.cmake for all of them? Most sport -g already, maybe that's good enough. But most have optimisations enabled nevertheless.

Yes, that's fine. The goal was to allow stepping-into code, right?

So it makes sense to to do the debugging-with-optimizations for all bundled deps, since we already do that for most of them.

That means we _don't_ need to check `if(CMAKE_BUILD_TYPE MATCHES Debug)` and we don't need to care about rebuilding `.deps` when the build-type changes.

mhinz committed:
third-party: build all deps with debugging symbols

When building with CMAKE_BUILD_TYPE=Debug, the dependencies are built like this:

| Dep        | Defaults  | Debug                                       |
|------------|-----------|---------------------------------------------|
| unibilium  | `-O2`     | `make CFLAGS=-O0 DEBUG=1`                   |
| msgpack    | `-g -O3`  | `cmake . -DCMAKE_C_FLAGS_DEBUG="-O0 -ggdb"` |
| libuv      | `-g -O2`  | `./configure CFLAGS="-O0 -ggdb"`            |
| luv        | `-g -O2`  | `cmake . -DCMAKE_C_FLAGS_DEBUG="-O0 -ggdb"` |
| libvterm   | not set   | `make CFLAGS=-O0 DEBUG=1`                   |
| libtermkey | not set   | `make CFLAGS=-O0 DEBUG=1`                   |
| jemalloc   | `-g3 -O3` | `./configure CFLAGS="-O0 -ggdb"`            |
| gperf      | `-g -O2`  | `./configure CXXFLAGS="-O0 -ggdb"`          |
| luajit     | `-g -O2`  | haven't checked yet                         |

This means that only unibilium, libtermkey, and libvterm don't build with
debugging symbols by default.

Build them with debugging symbols and optimisations that don't hinder
debugging: -Og -g

mhinz commented:
You can easily test this by breaking on functions that are all used at startup, e.g.

```
$ lldb build/bin/nvim
> b unibi_get_str
> b vterm_new
> b termkey_start
> r
> c
> c
> c
```

(It's probably similar/the same for gdb.) You'll get C sources for each of them, yay!

justinmk commented on third-party/cmake/BuildLibtermkey.cmake:
```diff
@@ -43,6 +43,7 @@ ExternalProject_Add(libtermkey
                               PREFIX=${DEPS_INSTALL_DIR}
                               PKG_CONFIG_PATH=${DEPS_LIB_DIR}/pkgconfig
                               CFLAGS=-fPIC
+                              CFLAGS+=-g
```
But this is not debug-with-optimizations, it's just debug. We set `HAS_OG_FLAG` in `CMakeLists.txt`, can that be used to add `-Og -g` if supported?

mhinz replied:
I'm confused. Are we talking about Debug or ReleaseWithDebInfo now?

justinmk replied:
The short answer is: just change this line to `CFLAGS+=-Og CFLAGS+=-g`.

The long answer is: `-Og` is not supported by some compilers/versions, so `-Og` should not be included. `HAS_OG_FLAG` checks whether the compiler supports `-Og`.

> Are we talking about Debug or ReleaseWithDebInfo now? What am I supposed to do for which build type?

I don't understand the question, we aren't doing anything different for different build-types.  We're just trying to enable optimizations if possible.

mhinz replied:
How about this?

```diff
diff --git i/third-party/CMakeLists.txt w/third-party/CMakeLists.txt
index ceebb4cf1..7404ca82c 100644
--- i/third-party/CMakeLists.txt
+++ w/third-party/CMakeLists.txt
@@ -4,0 +5,3 @@ project(NVIM_DEPS)
+# Used by check_c_compiler_flag()
+include(CheckCCompilerFlag)
+
@@ -13,0 +17,7 @@ endif()
+set(DEFAULT_C_FLAGS "-g -fPIC")
+
+check_c_compiler_flag(-Og HAS_OG_FLAG)
+if(HAS_OG_FLAG)
+  set(DEFAULT_C_FLAGS "-0g ${DEFAULT_C_FLAGS}")
+endif()
+
diff --git i/third-party/cmake/BuildLibtermkey.cmake w/third-party/cmake/BuildLibtermkey.cmake
index a99949e65..167692eee 100644
--- i/third-party/cmake/BuildLibtermkey.cmake
+++ w/third-party/cmake/BuildLibtermkey.cmake
@@ -45,2 +45 @@ ExternalProject_Add(libtermkey
-                              CFLAGS=-fPIC
-                              CFLAGS+=-g
+                              CFLAGS=${DEFAULT_C_FLAGS}
diff --git i/third-party/cmake/BuildLibvterm.cmake w/third-party/cmake/BuildLibvterm.cmake
index d6754f09b..76affb757 100644
--- i/third-party/cmake/BuildLibvterm.cmake
+++ w/third-party/cmake/BuildLibvterm.cmake
@@ -50,4 +50,3 @@ else()
-                                PREFIX=${DEPS_INSTALL_DIR}
-                                CFLAGS=-fPIC
-                                CFLAGS+=-g
-                                install)
+                                           PREFIX=${DEPS_INSTALL_DIR}
+                                           CFLAGS=${DEFAULT_C_FLAGS}
+                                           install)
diff --git i/third-party/cmake/BuildUnibilium.cmake w/third-party/cmake/BuildUnibilium.cmake
index 2b1084775..f56c1af8c 100644
--- i/third-party/cmake/BuildUnibilium.cmake
+++ w/third-party/cmake/BuildUnibilium.cmake
@@ -22,2 +22 @@ ExternalProject_Add(unibilium
-                            CFLAGS=-fPIC
-                            CFLAGS+=-g
+                            CFLAGS=${DEFAULT_C_FLAGS}
```

I just need to figure out a way how to quote arguments properly. Obviously the argument to `CFLAGS` here needs to be quoted. But.. easier said than done in cmake!

justinmk replied:
@mhinz isn't there a way to pass `HAS_OG_FLAG` from the top-level CMakeLists.txt to the `third-party/CMakeLists.txt` sub-project?

> how to quote arguments properly

AFAIK CMake tries to magically quote things on its own, so it's better to omit quotes than try to do them yourself.

mhinz replied:
Well, it's not a sub-project. They're two separate projects called by our Makefile. But [`load_cache()`](https://cmake.org/cmake/help/v3.0/command/load_cache.html) looks like it could be used instead.

The problem with the quoting above is that `CFLAGS=${DEFAULT_C_FLAGS}` leads to `make CFLAGS=-Og -g` which doesn't work. But `CFLAGS="${DEFAULT_C_FLAGS}`" leads to `make "CFLAGS=-Og -g`, which obviously doesn't work either. Cmake argument quoting is mayhem.

mhinz replied:
Another thing that just came to my mind.. with `load_cache()` we would reference another project and might include things we don't want. Yes, `load_cache()` provides a way to exclude entries, but it seems like a rabbit hole to me.

Moreover, deps are built before core, so after a `make clean` we would have no `HAS_OG_FLAG` set yet, so we would need to run the code anyway.

mhinz commented:
Is this good to go?
